### DISCUSSION TITLE
What is the intended interaction between root_path and source_dirs?

### DISCUSSION THREAD

**Discussion by cafhach** [fortran-lang/fortls#419] [Q&A]
I had issues with the `directories`-Option in VScode, i.e. the `source_dirs` argument in `fortls`. Reading into the code I cannot grasp what the intended behavior is. The [documentation](https://github.com/fortran-lang/fortls/blob/a23b93f0d21ecdf083230298cb0fd8b56f532021/docs/options.rst) says

    By default all directories under the current project will be recursively parsed for Fortran sources. Alternatively, one can define a series of directories for fortls to look for source files

From my perspective this implies, that `fortl` will respect my choice of `source_dirs`, if specified at all. But looking into method `serve_initialize` in https://github.com/fortran-lang/fortls/blob/a23b93f0d21ecdf083230298cb0fd8b56f532021/fortls/langserver.py I see that `self.root_path` is added to the source list in any case. Is this a bug or am I misinterpreting the docs? Thanks a lot!

**Comment by cafhach** [comment]
Method `_add_source_dirs` seems to throw away the value of `self.source_dirs` and replace it by a recursive search, starting with `self.root_path`. So the `source_ dirs` option seems to have no effect at all?

**Reply by gnikit** [reply]
I can't look at this right now but if no `root` options are sent in the LSP request, then the root is set to `""`. So when doing the glob resolution I am relatively certain that you end up in a situation where `resolve_globs(path, self.root_path)`  are both `""`. This is the area that needs investigation. Are these the args that go into the `resolve_globs`, if so what is the output of `resolve_globs`?

Zooming out, you can't spin up the server to parse sources without a root, I think that is rather obvious. It sounds to me that you should try some different configuration options.

In the future, please don't open bug reports for when you seek support. Either open a Q&A here or ask in Fortran-lang.

**Reply by cafhach** [reply]
Thanks for your reply and sorry for opening an issue.

I could look further into this and try to clean up the logic a bit but I don't know what the intended behavior is. Are there tests in place specifically to check Fortran file discovery and selection? I skimmed over the tests but didn't find anything like this.

**Reply by gnikit** [reply]
before trying to refactor I think it's important to understand how the server should act. The server **should** be run from a folder (root of the project). Anything else might run into problems. The general workflow with fortls is you allow it to scan your entire root for Fortran sources instead of manually specifying every single path. The source_dirs options is mostly used to extend the set of location to scan, not restrict them.
